src/proto/slp.rs

```diff
@@ -0,0 +1,198 @@
+                    online: 0,
+                    // FIXME(toqueteos): This is value read from server.properties file
+                    max: 20,
+                    sample: None
+                },
+                description: Description{
+                    text: "With custom favicons! Woot :D".to_string(),
+                },
+                favicon: format!("data:image/png;base64,{:}", favicon),
+            };
+            try!(StatusResponse { response: resp }.write(stream));
+            Ok(())
+        }
+        wrong_packet => Err(io::Error::new(InvalidInput, "invalid packet read", Some(format!("expecting C->S StatusRequest packet, got {:?}", wrong_packet))))
+    }
+}
+
+/// Server-side pong response, optional
+pub fn pong(mut stream: &mut TcpStream) -> io::Result<()> {
+    use packet::status::clientbound::Pong;
+    use packet::status::serverbound::Packet::{self, Ping};
+
+    // C->S: Ping packet
+    match try!(Packet::read(stream)) {
+        Ping(ping) => {
+            // S->C: Pong packet
+            try!(Pong { time: ping.time }.write(stream));
+            Ok(())
+        }
+        wrong_packet => Err(io::Error::new(InvalidInput, "invalid packet read", Some(format!("expecting C->S Ping packet, got {:?}", wrong_packet))))
+    }
+}
+
+/// Client-side Server List request
+pub fn request(mut stream: &mut TcpStream) -> io::Result<Response> {
+    use packet::status::serverbound::StatusRequest;
+    use packet::status::clientbound::Packet::{self, StatusResponse};
+
+    // C->S: Status Request packet
+    try!(StatusRequest.write(stream));
+
+    // S->C: Status Response packet
+    match try!(Packet::read(stream)) {
+        StatusResponse(resp) => Ok(resp.response),
+        wrong_packet => Err(io::Error::new(InvalidInput, "invalid packet read", Some(format!("expecting S->C StatusResponse packet, got {:?}", wrong_packet))))
+    }
+}
+
+/// Client-side ping request, optional
+pub fn ping(mut stream: &mut TcpStream) -> io::Result<i64> {
+    use packet::status::clientbound::Packet::{self, Pong};
+    use packet::status::serverbound::Ping;
+
+    // C->S: Ping packet
+    let start = time::get_time();
+    try!(Ping { time: start.sec }.write(stream));
+
+    // S->C: Pong packet
+    match try!(Packet::read(stream)) {
+        Pong(_) => {
+            let end = time::get_time();
+            let elapsed = end.sub(start).num_milliseconds();
+            println!("StatusPing elapsed={}", elapsed);
+            Ok(elapsed)
+        }
+        wrong_packet => Err(io::Error::new(InvalidInput, "invalid packet read", Some(format!("expecting S->C Pong packet, got {:?}", wrong_packet))))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::io::prelude::*;
+    use std::net::TcpStream;
+
+    use packet::handshake::Handshake;
+    use packet::{PacketWrite, NextState};
+
+    #[test]
+    #[cfg(vanilla_server_required)]
```

Comment:
This should do, now I need to know how to run `cargo test` with this cfg enabled.